amy/fiscal/base_views.py

```diff
@@ -0,0 +1,74 @@
+from functools import reduce
+from operator import add
+
+from django.shortcuts import get_object_or_404
+from django.views.generic import FormView
+
+from workshops.models import Membership
+
+
+class GetMembershipMixin:
+    def dispatch(self, request, *args, **kwargs):
+        self.membership = get_object_or_404(Membership, pk=self.kwargs["membership_id"])
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        kwargs["membership"] = self.membership
+        return super().get_context_data(**kwargs)
+
+
+class MembershipFormsetView(GetMembershipMixin, FormView):
+    template_name = "fiscal/membership_formset.html"
+
+    def get_formset_kwargs(self):
+        return {
+            "extra": 0,
+            "can_delete": True,
+        }
+
+    def get_form_class(self):
+        return self.get_formset(**self.get_formset_kwargs())
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["queryset"] = self.get_formset_queryset(self.membership)
+        return kwargs
+
+    def get_context_data(self, **kwargs):
+        kwargs["formset"] = self.get_form()
+        return super().get_context_data(**kwargs)
+
+    def form_valid(self, form):
+        form.save(commit=False)
+
+        if form.new_objects:
+            # assign membership to new instances
+            for instance in form.new_objects:
+                instance.membership = self.membership
+            form.model.objects.bulk_create(form.new_objects)
+
+        if form.changed_objects:
+            # Transpose changed objects to receive separately instances and
+            # changed data (fields changed for each instance). For example,
+            # changed_columns can look like:
+            #   (['role'], ['role', 'organization'], ['organization']).
+            # Out of this data we need ['role', 'organization'] to be passed to
+            # bulk_update().
+            changed_instances, changed_columns = zip(*form.changed_objects)
+            changed_columns_set = set(reduce(add, changed_columns))
+
+            # assign membership to changed instances
+            for instance in changed_instances:
+                instance.membership = self.membership
+            form.model.objects.bulk_update(changed_instances, changed_columns_set)
```

Comment:
Hmm this feels oddly heavy handed. When I'm boiling down what changes we're trying to add, it looks like the additional functionality is just something like: 

```py
 def form_valid(self, form):
        for instance in form.new_objects + form.changed_objects:
            instance.membership = self.membership
        return super().form_valid(form)
```

Would something like that not work? And if not and we do need to enumerate things, just to ensure I'm understanding does the optimization get us anything? Are we doing the operations on potentially large numbers of forms?

Reply:
What you have looks correct. And you’re comment explains things well. I do think it’s a little less readable so if the optimization isn’t needed it might be better to just go with your original example for now.

Reply:
I agree that this is way less readable than previous version, even if it has better performance. I'll remove the last commit which introduced the changes.